airframe-sql/src/main/scala/wvlet/airframe/sql/model/Expression.scala

```diff
@@ -52,6 +52,50 @@ sealed trait Expression extends TreeNode[Expression] with Product {
     primaryConstructor.newInstance(args.toArray[AnyRef]: _*).asInstanceOf[Expression]
   }
 
+  def transformPlan(rule: PartialFunction[LogicalPlan, LogicalPlan]): Expression = {
+    def recursiveTransform(arg: Any): AnyRef = {
+      arg match {
+        case e: Expression  => e.transformPlan(rule)
+        case l: LogicalPlan => l.transform(rule)
+        case Some(x)        => Some(recursiveTransform(x))
+        case s: Seq[_]      => s.map(recursiveTransform _)
+        case other: AnyRef  => other
+        case null           => null
+      }
+    }
+
+    val newArgs = productIterator.map(recursiveTransform)
+    createInstance(newArgs)
+  }
+
+  def traversePlan[U](rule: PartialFunction[LogicalPlan, U]): Unit = {
```

Comment:
Simply changing `LogicalPlan.traverse()` to accept `PartialFunction[TreeNode[_], U]` works, but if we do so, `traverse()` will be able to handle `Expression`, not only `LogicalPlan` although `LogicalPlan` already have `traverseExpressions()`.

I think adding `traversePlan()` is more consistent in terms of API design unless we drop methods for `Expression`.